src/protocols/multi_party_ecdsa/gg_2020/state_machine/keygen.rs

```diff
@@ -0,0 +1,507 @@
+//! High-level keygen protocol implementation
+
+use std::fmt;
+use std::mem::replace;
+use std::time::Duration;
+
+use curv::cryptographic_primitives::proofs::sigma_dlog::DLogProof;
+use curv::cryptographic_primitives::secret_sharing::feldman_vss::VerifiableSS;
+use curv::elliptic::curves::secp256_k1::{FE, GE};
+use round_based::containers::{
+    push::{Push, PushExt},
+    *,
+};
+use round_based::{IsCritical, Msg, StateMachine};
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+use crate::protocols::multi_party_ecdsa::gg_2020;
+
+mod rounds;
+
+use private::InternalError;
+pub use rounds::{LocalKey, ProceedError};
+use rounds::{Round0, Round1, Round2, Round3, Round4};
+
+/// Keygen protocol state machine
+///
+/// Successfully completed keygen protocol produces [LocalKey] that can be used in further
+/// [signing](super::sign) protocol.
+pub struct Keygen {
+    round: R,
+
+    msgs1: Option<Store<BroadcastMsgs<gg_2020::party_i::KeyGenBroadcastMessage1>>>,
+    msgs2: Option<Store<BroadcastMsgs<gg_2020::party_i::KeyGenDecommitMessage1>>>,
+    msgs3: Option<Store<P2PMsgs<(VerifiableSS<GE>, FE)>>>,
+    msgs4: Option<Store<BroadcastMsgs<DLogProof<GE>>>>,
+
+    msgs_queue: Vec<Msg<ProtocolMessage>>,
+
+    party_i: u16,
+    party_n: u16,
+}
+
+impl Keygen {
+    /// Constructs a party of keygen protocol
+    ///
+    /// Takes party index `i` (in range `[1; n]`), threshold value `t`, and total number of
+    /// parties `n`. Party index identifies this party in the protocol, so it must be guaranteed
+    /// to be unique.
+    ///
+    /// Returns error if:
+    /// * `n` is less than 2, returns [Error::TooFewParties]
+    /// * `t` is not in range `[1; n-1]`, returns [Error::InvalidThreshold]
+    /// * `i` is not in range `[1; n]`, returns [Error::InvalidPartyIndex]
+    pub fn new(i: u16, t: u16, n: u16) -> Result<Self> {
+        if n < 2 {
+            return Err(Error::TooFewParties);
+        }
+        if t == 0 || t >= n {
+            return Err(Error::InvalidThreshold);
+        }
+        if i == 0 || i > n {
+            return Err(Error::InvalidPartyIndex);
+        }
+        let mut state = Self {
+            round: R::Round0(Round0 { party_i: i, t, n }),
+
+            msgs1: Some(Round1::expects_messages(i, n)),
+            msgs2: Some(Round2::expects_messages(i, n)),
+            msgs3: Some(Round3::expects_messages(i, n)),
+            msgs4: Some(Round4::expects_messages(i, n)),
+
+            msgs_queue: vec![],
+
+            party_i: i,
+            party_n: n,
+        };
+
+        state.proceed_round(false)?;
+        Ok(state)
+    }
+
+    fn gmap_queue<'a, T, F>(&'a mut self, mut f: F) -> impl Push<Msg<T>> + 'a
```

Comment:
Keygen state machine holds a [msgs_queue] that's `Vec<Msg<ProtocolMessage>>`. It's a queue of outgoing messages, [rounds] extend it when they want to send something to other parties. `Vec<Msg<ProtocolMessage>>` implements `Push<Msg<ProtocolMessage>>` (`ProtocolMessage`, roughly saying, is an enum over all messages that appear int the protocol), whereas rounds expect it to implement `Push<Msg<M>>` (`M` - round-specific message, e.g. [KeyGenBroadcastMessage1]).

So gmap takes closure f `M -> ProtocolMessage`, a queue `impl Push<Msg<ProtocolMessage>>`, and returns `impl Push<Msg<M>>` for round-specific messages. I call it gmap because it has signature `(a -> b) -> f b -> f a` that's contravariant of map. The name I just hoogled:

<img width="755" alt="image" src="https://user-images.githubusercontent.com/14890036/123589313-3e87d680-d813-11eb-9b01-fb192495b676.png">

[msgs_queue]: https://github.com/ZenGo-X/multi-party-ecdsa/blob/3ed044d67c7d454ba4d8609c151ed9de14fbdf4b/src/protocols/multi_party_ecdsa/gg_2020/state_machine/keygen.rs#L38
[rounds]: https://github.com/ZenGo-X/multi-party-ecdsa/blob/exp_state_machine/src/protocols/multi_party_ecdsa/gg_2020/state_machine/keygen/rounds.rs
[KeyGenBroadcastMessage1]: https://github.com/ZenGo-X/multi-party-ecdsa/blob/3ed044d67c7d454ba4d8609c151ed9de14fbdf4b/src/protocols/multi_party_ecdsa/gg_2020/state_machine/keygen/rounds.rs#L24

In signing implementation, I managed to bypass need of gmap by wrapping queue and implementing Push<M> for it: see [here][bypassing]. However, it required me to write a macro (or I would have to write much more code). I have an option to reuse this macro for keygen as well, but probably this macro should rather be moved to `round_based` crate. So I'm thinking what's a better way to do this.

[bypassing]: https://github.com/ZenGo-X/multi-party-ecdsa/blob/3ed044d67c7d454ba4d8609c151ed9de14fbdf4b/src/protocols/multi_party_ecdsa/gg_2020/state_machine/sign.rs#L563-L589